// objection/lib/queryBuilder/operations/PatchGraphOperation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _RelationExpression = require('../RelationExpression');

var _RelationExpression2 = _interopRequireDefault(_RelationExpression);

var _DelegateOperation2 = require('./DelegateOperation');

var _DelegateOperation3 = _interopRequireDefault(_DelegateOperation2);

var _UpdateOperation = require('./UpdateOperation');

var _UpdateOperation2 = _interopRequireDefault(_UpdateOperation);

var _inserter = require('../graphInserter/inserter');

var _inserter2 = _interopRequireDefault(_inserter);

var _GraphInserter = require('../graphInserter/GraphInserter');

var _GraphInserter2 = _interopRequireDefault(_GraphInserter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PatchGraphOperation = function (_DelegateOperation) {
  (0, _inherits3.default)(PatchGraphOperation, _DelegateOperation);

  function PatchGraphOperation(name, opt) {
    (0, _classCallCheck3.default)(this, PatchGraphOperation);

    var _this = (0, _possibleConstructorReturn3.default)(this, _DelegateOperation.call(this, name, opt));

    if (!_this.delegate.is(_UpdateOperation2.default)) {
      throw new Error('Invalid delegate');
    }

    // Our delegate operation inherits from `UpdateOperation`. Disable the call-time
    // validation. We do the validation in onAfterQuery instead.
    _this.delegate.modelOptions.skipValidation = true;
    return _this;
  }

  PatchGraphOperation.prototype.call = function call(builder, args) {
    var retVal = _DelegateOperation.prototype.call.call(this, builder, args);

    // We resolve this query here and will not execute it. This is because the root
    // value may depend on other models in the graph and cannot be inserted first.
    builder.resolve([]);

    return retVal;
  };

  PatchGraphOperation.prototype.onBeforeInternal = function onBeforeInternal() {
    // Do nothing. We override this with empty implementation so that
    // the $beforeUpdate() hooks are not called twice for the root models.
  };

  PatchGraphOperation.prototype.onBuild = function onBuild() {}
  // Do nothing.


  // We overrode all other hooks but this one and do all the work in here.
  // This is a bit hacky.
  ;

  PatchGraphOperation.prototype.onAfterQuery = function onAfterQuery(builder) {
    var eager = _RelationExpression2.default.fromGraph(this.model);

    return builder.modelClass().query().childQueryOf(builder).copyFrom(builder, /where/).eager(eager).then(function (models) {
      var fetched = void 0;

      if (Array.isArray(models)) {
        if (models.length > 1) {
          throw new Error('A query with `patchGraph` resulted in multiple rows. `patchGraph` can only be applied to queries that result in a single row.');
        }

        fetched = models[0];
      } else {
        fetched = models;
      }

      console.log(models);
    });
  };

  PatchGraphOperation.prototype.onAfterInternal = function onAfterInternal() {
    return this.model;
  };

  (0, _createClass3.default)(PatchGraphOperation, [{
    key: 'model',
    get: function get() {
      return this.delegate.model;
    }
  }]);
  return PatchGraphOperation;
}(_DelegateOperation3.default);

exports.default = PatchGraphOperation;
